<a href="https://colab.research.google.com/github/salmazayaf/CaseStudy-Data-Wrangling/blob/main/Group_F_CaseStudy_Data_Wrangling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<center><h1><strong><font color="blue">Studi Kasus Data Wrangling</font></strong></h1></center>

![taudata Analytics](https://github.com/taudataid/eLearning/raw/master/images/covers/cover_ui-n-taudata.jpg "taudata Analytics")

## **Group F**


*   Esly Vani Frasanti Saragih - 2106700782
*   Adriaan Aylsworth Waleleng - 2106725072
*   Lutfia Maulidina - 2006483151
*   Salma Mazaya Fasya - 2006536126
*   Zahra Novanka - 1806185834
*   Yumna Izzati - 2106633191

In [ ]:
import warnings; warnings.simplefilter('ignore')
import pandas as pd

try: # Running Locally, yakinkan "file_" berada di folder "data"
    mengajar = pd.read_csv('data/mengajar.csv', encoding='utf8')
    dosen = pd.read_csv('data/tb_dosen.csv', encoding='utf8')
    kelas = pd.read_csv('data/tb_kelas.csv', encoding='utf8')
    krs = pd.read_csv('data/tb_krs.csv', encoding='utf8')
    mahasiswa = pd.read_csv('data/tb_mahasiswa.csv', encoding='utf8')
    matkul = pd.read_csv('data/tb_mata_kuliah.csv', encoding='utf8')
    nilai = pd.read_csv('data/tb_nilai.csv', encoding='utf8')
except: # Running in Google Colab
    !mkdir data
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_nilai.csv
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/mengajar.csv
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_dosen.csv
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_kelas.csv
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_krs.csv
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_mahasiswa.csv
    !wget -P data/ https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_mata_kuliah.csv
    mengajar = pd.read_csv('data/mengajar.csv', encoding='utf8')
    dosen = pd.read_csv('data/tb_dosen.csv', encoding='utf8')
    kelas = pd.read_csv('data/tb_kelas.csv', encoding='utf8')
    krs = pd.read_csv('data/tb_krs.csv', encoding='utf8')
    mahasiswa = pd.read_csv('data/tb_mahasiswa.csv', encoding='utf8')
    matkul = pd.read_csv('data/tb_mata_kuliah.csv', encoding='utf8')
    nilai = pd.read_csv('data/tb_nilai.csv', encoding='utf8')

mengajar.shape, dosen.shape, kelas.shape, krs.shape, mahasiswa.shape, matkul.shape, nilai.shape

--2022-11-18 03:30:17--  https://raw.githubusercontent.com/taudataid/eLearning/master/data/tb_nilai.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.111.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1136 (1.1K) [text/plain]
Saving to: ‘data/tb_nilai.csv’

tb_nilai.csv        100%[===================>]   1.11K  --.-KB/s    in 0s      

2022-11-18 03:30:17 (69.7 MB/s) - ‘data/tb_nilai.csv’ saved [1136/1136]

--2022-11-18 03:30:17--  https://raw.githubusercontent.com/taudataid/eLearning/master/data/mengajar.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 329 [text/plain]
Savi

((14, 4), (4, 6), (4, 4), (7, 6), (15, 8), (9, 3), (33, 7))

#NOMOR 1

In [ ]:
# Code untuk Soal No 01 diberikan sebagai Contoh dan Bonus
# Pelajari dengan baik

nilai_dosen = nilai.groupby(nilai['NIDN']).mean()[['nilai']]
nilai_dosen = nilai_dosen.rename({'nilai': 'mean_nilai'}, axis=1) 
nilai_dosen = pd.merge(dosen, nilai_dosen, left_on='NIDN',  right_index=True)
nilai_dosen = nilai_dosen[['NIDN', 'nama', 'mean_nilai']]
sb = nilai.groupby(nilai['NIDN']).std()[['nilai']]
sb = sb.rename({'nilai': 'std_nilai'}, axis=1) 
nilai_dosen = nilai_dosen.set_index(['NIDN']) 
nilai_dosen = pd.concat([nilai_dosen, sb], axis=1)
nilai_dosen

## Interpretasi 01: 

Dapat dilihat dari tabel diatas bahwa dosen yang rata-rata nilai mahasiswanya paling tinggi adalah Andri Basuki dengan rata-rata 85,75. Dilihat dari standar deviasi juga memiliki nilai yang paling kecil diantara dosen-dosen lain sehingga mahasiswa-mahasiswa yang diampunya memiliki jarak nilai yang tidak jauh satu sama lain. Nilai mean cukup merepresentasikan data nilainya.  

#NOMOR 2

In [ ]:
# Code Untuk Soal No 02

nilai_dosen2 = nilai.groupby(nilai['NIDN']).mean()[['nilai']]
nilai_dosen2 = nilai_dosen2.rename({'nilai': 'mean_nilai'}, axis=1) 
nilai_dosen2 = pd.merge(dosen, nilai_dosen2, left_on='NIDN', right_index=True)
nilai_dosen2 = nilai_dosen2[['NIDN', 'nama','Pendidikan_Terakhir','mean_nilai']]
sb = nilai.groupby(nilai['NIDN']).std()[['nilai']]
sb = sb.rename({'nilai': 'std_nilai'}, axis=1) 
nilai_dosen2 = nilai_dosen2.set_index(['NIDN']) 
nilai_dosen2 = pd.concat([nilai_dosen2, sb], axis=1) 
nilai_dosen2 = nilai_dosen2.set_index('Pendidikan_Terakhir')
nilai_dosen2

nama  mean_nilai  std_nilai
Pendidikan_Terakhir                                      
S3                    Andri Basuki   85.750000   7.741355
S3                        zulfikar   80.428571  15.883504
S2                   Annisa Fitria   60.090909  16.220077
S2                            yudi   70.571429  13.010984

## Interpretasi 02:

Berdasarkan tabel diatas, jika mengurutkan rata-rata nilai yang diberikan dosen dengan pendidikan terakhirnya, Dosen bernama Andi Basuki dan zulfikar dari empat dosen merupakan dua dosen yang memiliki rata-rata nilai yang lebih tinggi dibanding kedua dosen lainnya. Andi basuki dan zulfikar juga berpendidikan terakhir S3 sedangkan kedua dosen lainnya berpendidikan terakhir S2. Jadi, dapat disimpulkan bahwa dari data yang tersedia, dosen dengan pendidikan terakhir S3 menghasilkan mahasiswa dengan nilai yang lebih baik.

#NOMOR 3

In [ ]:
# Code Soal No 03



## Interpretasi 03:

* 

#NOMOR 4

In [ ]:
Gender = mahasiswa[['nim','nama','Gender']]
nilai = nilai[['nim','nilai']]
tabel = pd.merge(Gender,nilai, on ='nim', how ='left')
tabel

nim            nama  Gender  nilai
0   194015   Muhammad Amir    Pria     70
1   194015   Muhammad Amir    Pria     40
2   194001   Samsul Kurnia    Pria     80
3   194001   Samsul Kurnia    Pria     55
4   194002  Bambang Arifin    Pria     85
5   194002  Bambang Arifin    Pria     50
6   194002  Bambang Arifin    Pria     60
7   194003            Udin    Pria     90
8   194004            Iwan    Pria     92
9   194004            Iwan    Pria     90
10  194004            Iwan    Pria     56
11  194004            Iwan    Pria     65
12  194005  Bambang Arifin    Pria     91
13  194005  Bambang Arifin    Pria     92
14  194006            Wati  Wanita     93
15  194006            Wati  Wanita     85
16  194007             Ani  Wanita     85
17  194007             Ani  Wanita     55
18  194007             Ani  Wanita     70
19  194007             Ani  Wanita     74
20  194008       Rahmawati  Wanita     60
21  194008       Rahmawati  Wanita     60
22  194008       Rahmawati  Wanita     72
23  194009            Asep    Pria     63
24  194009            Asep    Pria     91
25  194010            Rina  Wanita     62
26  194010            Rina  Wanita     90
27  194011            Sari  Wanita     61
28  194011            Sari  Wanita     31
29  194012            Bowo    Pria     70
30  194013            Budi    Pria     90
31  194014           Putri  Wanita     88
32  194014           Putri  Wanita     88

In [ ]:
Gender = mahasiswa[['nim','Gender']]
nilai = nilai[['nim','nilai']]
tabel = pd.merge(Gender,nilai, on ='nim', how ='left')
tabel.groupby('Gender')[['nilai']].agg('mean')

nilai
Gender           
Pria    73.888889
Wanita  71.600000

## Interpretasi 04

Dari tabel di atas, terbukti benar adanya perbedaan kemampuan akademik antara mahasiswa dan mahasiswi. Dilihat dari mean-nya, nilai mahasiswa rata-rata lebih tinggi dibandingkan mahasiswi. Sehingga dapat disimpulkan, mahasiswa memiliki kemampuan akademik lebih baik dibandingkan mahasiswi. 

#NOMOR 5

In [ ]:
nilai_angkatan = nilai.groupby(nilai['nim']).mean()[['nilai']]
nilai_angkatan = nilai_angkatan.rename({'nilai': 'mean_nilai'}, axis=1)
nilai_angkatan = pd.merge(mahasiswa, nilai_angkatan, left_on = 'nim', right_index = True)
nilai_angkatan = nilai_angkatan[['Angkatan', 'mean_nilai', 'nim']]
sb = nilai.groupby(nilai['nim']).std()[['nilai']]
sb = sb.rename({'nilai': 'std_nilai'}, axis=1) 
nilai_angkatan = nilai_angkatan.set_index(['nim']) 
nilai_angkatan = pd.concat([nilai_angkatan, sb], axis=1) 
nilai_angkatan = nilai_angkatan.pivot_table(index='Angkatan')

nilai_angkatan

mean_nilai  std_nilai
Angkatan                       
2019        75.59375  13.385601
2020        73.00000  13.547877

## Interpretasi 05

Berdasarkan hasil pengolahan data, terlihat bahwa ada perbedaan prestasi akademik antar angkatan mahasiswa. Pada tabel hasil di atas, mahasiswa angkatan 2019 memiliki perolehan nilai rata-rata sebesar 75.59375 dengan standar deviasi 13.385601. Sementara itu, mahasiswa angkatan 2020 memiliki perolehan nilai rata-rata sebesar 73.00000 dengan standar deviasi 13.547877 yang artinya jarak nilai antar mahasiswa 2020 lebih jauh daripada angkatan 2019.
Dengan perolehan nilai rata-rata lebih tinggi dan standar deviasi yang lebih kecil, dapat disimpulkan bahwa prestasi akademik mahasiswa angkatan 2019 lebih baik dibandingkan mahasiswa angkatan 2020.

#NOMOR 6

In [ ]:
#Drop Data unwanted.
#Digunakan Drop Data untuk menghapus data yang mengandung missing value pada data ataupun menghapus variabel yang memiliki banyak missing value.

mengajar.drop(mengajar.columns[14:26], axis=1, inplace=True)
dosen.drop(dosen.columns[14:26], axis=1, inplace=True)
kelas.drop(kelas.columns[14:26], axis=1, inplace=True)
krs.drop(krs.columns[14:26], axis=1, inplace=True)
mahasiswa.drop(mahasiswa.columns[14:26], axis=1, inplace=True)
matkul.drop(matkul.columns[14:26], axis=1, inplace=True)
nilai.drop(nilai.columns[14:26], axis=1, inplace=True)

print(mengajar.shape, dosen.shape, kelas.shape, krs.shape, mahasiswa.shape, matkul.shape, nilai.shape)

(14, 4) (4, 6) (4, 4) (7, 6) (15, 8) (9, 3) (33, 7)


In [ ]:
krs

id id_mata_kuliah kode_kelas     nim  tahun semester
0   1          KAL01      SIS1J  193001   2021    genap
1   2          STA01       D101  193001   2021    genap
2   3           OR01       D212  193001   2021    genap
3   4          MK002       P601  194001   2021    genap
4   5          MK002       P601  194001   2021    genap
5   6          MK003      SIS1J  194001   2021    genap
6   7          MK004       D101  194001   2021    genap

In [ ]:
nilai_angkatan = nilai.groupby(nilai['nim']).mean()[['nilai']]
nilai_angkatan = nilai_angkatan.rename({'nilai': 'mean_nilai'}, axis=1) 
nilai_angkatan = pd.merge(mahasiswa, nilai_angkatan, left_on='nim', right_index=True)
nilai_angkatan = nilai_angkatan[['nim', 'Asal_Daerah', 'mean_nilai']]
sb_1 = nilai.groupby(nilai['nim']).std()[['nilai']]
sb_1 = sb_1.rename({'nilai': 'std_nilai'}, axis=1) 
nilai_angkatan = nilai_angkatan.set_index(['nim']) 
nilai_angkatan = pd.concat([nilai_angkatan, sb_1], axis=1)
nilai_angkatan.pivot_table(index='Asal_Daerah')

mean_nilai  std_nilai
Asal_Daerah                       
Lokal         78.000000  10.283147
Luar Daerah   68.958333  17.891313

## Interpretasi 06

*  Dari data-data yang digunakan di atas (yaitu mengajar, dosen, kelas, krs, mahasiswa, matkul, dan nilai), terdapat duplikat data pada data krs di kolom id nomor 4 dan 5. Maka data dari krs ini belum dapat digunakan secara baik.

* Dari pembaharuan soal no.3 di dapat ide untuk membandingkan perbedaan performa akademik antara mahasiswa yang berasal dari Lokal (dalam daerah) dengan Luar Daerah. Dan dapat diketahui dari hasil yang diperoleh pada tabel di atas bahwa performa akademik mahasiswa lokal memiliki nilai rata-rata yang lebih tinggi dibandingkan dengan yang di luar daerah. Kemudian juga terdapat perbedaan yang signifikan antara nilai standar deviasi nya dari asal daerah lokal dan luar daerah.

#TABEL-TABEL  

In [ ]:
mengajar

NIDN kode_MK Semester  Tahun
0    123   KAL01   ganjil   2020
1    123   KAL01   ganjil   2019
2    123   STA01    genap   2020
3    123   MAT01    genap   2020
4    123   STA02   ganjil   2019
5    234    OR01   ganjil   2019
6    234    CS01    genap   2020
7    787   KAL01   ganjil   2021
8    787   KAL01   ganjil   2022
9    787   STA02   ganjil   2021
10   345   MAT73    genap   2020
11   345   MAT53    genap   2019
12   345   STA02   ganjil   2019
13   345   KAL02   ganjil   2020

In [ ]:
dosen

NIDN           nama             email Pendidikan_Terakhir  Telephone  \
0   123   Andri Basuki   anbr@googol.com                  S3   11230000   
1   234       zulfikar   dosen@email.com                  S3   22340000   
2   787           yudi    yudi@email.com                  S2   55670000   
3   345  Annisa Fitria  annniza@yyyy.com                  S2   88970000   

  Jenis_Kelamin  
0          Pria  
1          Pria  
2          Pria  
3        Wanita

In [ ]:
kelas

kode_kelas nama_kelas lokasi_kelas  Kapasitas
0      SIS1J  Reguler A     Gedung D         10
1       D101  Reguler A     Gedung D         15
2       D212  Reguler B     Gedung D         40
3       P601  Reguler B   Gedung PLT         30

In [ ]:
krs

id id_mata_kuliah kode_kelas     nim  tahun semester
0   1          KAL01      SIS1J  193001   2021    genap
1   2          STA01       D101  193001   2021    genap
2   3           OR01       D212  193001   2021    genap
3   4          MK002       P601  194001   2021    genap
4   5          MK002       P601  194001   2021    genap
5   6          MK003      SIS1J  194001   2021    genap
6   7          MK004       D101  194001   2021    genap

In [ ]:
mahasiswa

nim            nama                    email  Gender Asal_Sekolah  \
0   194015   Muhammad Amir       Samarinda@inet.com    Pria          SMU   
1   194001   Samsul Kurnia        Bontang@email.com    Pria          SMU   
2   194002  Bambang Arifin         Sangata@inet.com    Pria          SMU   
3   194003            Udin            Udin@inet.com    Pria           MA   
4   194004            Iwan            Iwan@inet.com    Pria           MA   
5   194005  Bambang Arifin  Bambang Arifin@inet.com    Pria          SMK   
6   194006            Wati            Wati@inet.com  Wanita          SMK   
7   194007             Ani             Ani@inet.com  Wanita          SMK   
8   194008       Rahmawati       Rahmawati@inet.com  Wanita          SMU   
9   194009            Asep            Asep@inet.com    Pria          SMU   
10  194010            Rina            rina@inet.com  Wanita           MA   
11  194011            Sari            sari@inet.com  Wanita           MA   
12  194012            Bowo            Bowo@inet.com    Pria          SMU   
13  194013            Budi            Budi@inet.com    Pria          SMK   
14  194014           Putri           putri@inet.com  Wanita          SMK   

    Telephone  Angkatan  Asal_Daerah  
0    12345678      2019  Luar Daerah  
1    12345679      2019        Lokal  
2    12345680      2019  Luar Daerah  
3    12345681      2019        Lokal  
4    12345682      2019  Luar Daerah  
5    12345683      2019        Lokal  
6    12345684      2019        Lokal  
7    12345685      2019  Luar Daerah  
8    12345686      2020        Lokal  
9    12345687      2020  Luar Daerah  
10   12345688      2020        Lokal  
11   12345689      2020        Lokal  
12   12345690      2020  Luar Daerah  
13   12345691      2020        Lokal  
14   12345692      2020        Lokal

In [ ]:
matkul

id_mata_kuliah           nama_mata_kuliah  SKS
0          KAL01             Kalkulus Dasar    3
1          STA01       Statistika Elementer    3
2          MAT01           Matematika Dasar    3
3          STA02      Statistika Matematika    4
4           OR01         Operation Research    4
5           CS01  Pendahuluan Ilmu Komputer    3
6          MAT73            Aljabar Abstrak    4
7          MAT53             Aljabar linear    3
8          KAL02            Kalkulus Lanjut    4

In [ ]:
nilai

nim id_mata_kuliah  NIDN  nilai Semester  Tahun     rand
0   194015          KAL01   123     70   ganjil   2019  0.50463
1   194001          KAL01   123     80   ganjil   2019      NaN
2   194002          KAL01   123     85   ganjil   2019      NaN
3   194003          KAL01   123     90   ganjil   2019      NaN
4   194004          KAL01   123     92   ganjil   2019      NaN
5   194005          KAL01   123     91   ganjil   2019      NaN
6   194006          KAL01   123     93   ganjil   2019      NaN
7   194007          KAL01   123     85   ganjil   2019      NaN
8   194008          KAL01   787     60   ganjil   2020      NaN
9   194009          KAL01   787     63   ganjil   2020      NaN
10  194010          KAL01   787     62   ganjil   2020      NaN
11  194011          KAL01   787     61   ganjil   2020      NaN
12  194012          KAL01   787     70   ganjil   2020      NaN
13  194013          KAL01   787     90   ganjil   2020      NaN
14  194014          KAL01   787     88   ganjil   2020      NaN
15  194004           CS01   234     90    genap   2020      NaN
16  194005           CS01   234     92    genap   2020      NaN
17  194006           CS01   234     85    genap   2020      NaN
18  194007           CS01   234     55    genap   2020      NaN
19  194008           CS01   234     60    genap   2020      NaN
20  194009           CS01   234     91    genap   2020      NaN
21  194010           CS01   234     90    genap   2020      NaN
22  194015          STA02   345     40   ganjil   2021      NaN
23  194002          STA02   345     50   ganjil   2021      NaN
24  194004          STA02   345     56   ganjil   2021      NaN
25  194007          STA02   345     70   ganjil   2021      NaN
26  194008          STA02   345     72   ganjil   2021      NaN
27  194011          STA02   345     31   ganjil   2021      NaN
28  194014          STA02   345     88   ganjil   2021      NaN
29  194001          KAL02   345     55   ganjil   2020      NaN
30  194002          KAL02   345     60   ganjil   2020      NaN
31  194004          KAL02   345     65   ganjil   2020      NaN
32  194007          KAL02   345     74   ganjil   2020      NaN